File: 22nd-April-Smaller_Sum/c++/solution.cpp
/*
  This code is written by Bhaskar
*/

class Solution{
public:
    vector<long long> smallerSum(int n,vector<int> &arr){
    vector<int> list(arr);
    sort(list.begin(), list.end()); 

    unordered_map<int, long long> hm;
    long long sum = 0;

    for(int i = 0; i < n; i++) {
        if(!hm.count(list[i])) {
            hm[list[i]] = sum; 
        }
        sum += list[i];
    }

    vector<long long> ans(n);

    for(int i = 0; i < n; i++) {
        ans[i] = hm[arr[i]];
    }

    return ans; 
}
};

File: 22nd-April-Smaller_Sum/java/solution.java
/*
  This code is written by Bhaskar
*/

//User function Template for Java
class Solution
{
    public long[] smallerSum(int n,int arr[])
    {
        ArrayList<Integer> list = new ArrayList<>();
        for(int x : arr){
            list.add(x);
        }
        Collections.sort(list);

        HashMap<Integer , Long > hm = new HashMap<>();

        long sum = 0;

        for(int i=0 ; i < n ; i++){
            if(!hm.containsKey(list.get(i))){
                hm.put(list.get(i) , sum);

            }

            sum += list.get(i);
        }

        long[] ans = new long[n];

        for(int i=0 ; i < n ; i++){
            ans[i] = hm.get(arr[i]);
        }

        return ans;

    }
}

